ffi/src/primitives.rs

```diff
@@ -169,16 +171,43 @@ impl<'itm> IntoFfiSliceRef<'itm> for core::cmp::Ordering {
     }
 }
 
+/// Trait for replacing unsupported types with supported ones when crossing FFI-boundary (for example in case of wasm)
+pub trait PrimitiveRepr {
+    /// Type used to represent [`Self`] when crossing FFI-boundry
+    type Repr;
+}
+
+macro_rules! primitive_repr_impls {
+    ($($source:ty => $target:ty),+ $(,)?) => {$(
+        impl PrimitiveRepr for $source {
+            type Repr = $target;
+        }
+    )+};
+}
+macro_rules! primitive_repr_self_impls {
+    ($($source:ty),+ $(,)?) => {
+        primitive_repr_impls! {
+            $($source => $source),*
+        }
+    };
+}
+
+#[cfg(feature = "wasm")]
+primitive_repr_impls! {u8 => u32, u16 => u32, i8 => i32, i16 => i32}
+#[cfg(not(feature = "wasm"))]
+primitive_repr_self_impls! {u8, u16, i8, i16}
+primitive_repr_self_impls! {u32, u64, u128, i32, i64, i128}
+
 macro_rules! primitive_impls {
     ( $( $ty:ty ),+ $(,)? ) => { $(
         unsafe impl ReprC for $ty {}
 
         impl TryFromReprC<'_> for $ty {
-            type Source = Self;
+            type Source = <$ty as PrimitiveRepr>::Repr;
```

Comment:
```suggestion
            #[cfg(feature = "wasm")]
            type Source = <$ty as WasmRepr>::Repr;
            #[cfg(not(feature = "wasm"))]
            type Source = Self;
```

here is what I suggest. `try_into` will work in both cases and compiler should elide `try_into` when `Source = Self`